proxy/envoy/discovery.go

```diff
@@ -125,101 +173,178 @@ func (ds *DiscoveryService) Run() {
 	}
 }
 
+// GetCacheStats returns the statistics for cached discovery responses.
+func (ds *DiscoveryService) GetCacheStats(_ *restful.Request, response *restful.Response) {
+	ds.cacheMu.Lock()
+	defer ds.cacheMu.Unlock()
+	if err := response.WriteEntity(ds.cacheStats); err != nil {
+		glog.Warning(err)
+	}
+}
+
+func (ds *DiscoveryService) clearCache() {
+	ds.cacheMu.Lock()
+	defer ds.cacheMu.Unlock()
+	glog.Infof("Cleared discovery service cache")
+	ds.cache = make(map[string][]byte)
+}
+
+// caller must hold cache mutex lock
+func (ds *DiscoveryService) cacheHitMiss(key string, hit bool) {
+	if ds.disableCache {
+		return
+	}
+	s, ok := ds.cacheStats.Stats[key]
+	if !ok {
+		s = &discoveryCacheStatEntry{}
+		ds.cacheStats.Stats[key] = s
+	}
+	if hit {
+		s.Hit++
+	} else {
+		s.Miss++
+	}
+}
+
 // ListEndpoints responds to SDS requests
 func (ds *DiscoveryService) ListEndpoints(request *restful.Request, response *restful.Response) {
-	hostname, ports, tags := model.ParseServiceKey(request.PathParameter(ServiceKey))
-	// envoy expects an empty array if no hosts are available
-	out := make([]*host, 0)
-	for _, ep := range ds.services.Instances(hostname, ports.GetNames(), tags) {
-		out = append(out, &host{
-			Address: ep.Endpoint.Address,
-			Port:    ep.Endpoint.Port,
-		})
-	}
-	if err := response.WriteEntity(hosts{Hosts: out}); err != nil {
-		glog.Warning(err)
+	ds.cacheMu.Lock()
+	defer ds.cacheMu.Unlock()
+
+	key := request.Request.URL.String()
+	out, cached := ds.cache[key]
+	if ds.disableCache || !cached {
+		hostname, ports, tags := model.ParseServiceKey(request.PathParameter(ServiceKey))
+		// envoy expects an empty array if no hosts are available
+		hostArray := make([]*host, 0)
+		for _, ep := range ds.services.Instances(hostname, ports.GetNames(), tags) {
+			hostArray = append(hostArray, &host{
+				Address: ep.Endpoint.Address,
+				Port:    ep.Endpoint.Port,
+			})
+		}
+		var err error
+		if out, err = json.MarshalIndent(hosts{Hosts: hostArray}, " ", " "); err != nil {
+			errorResponse(response, http.StatusInternalServerError, err.Error())
+			return
+		}
+		ds.cache[key] = out
```

Comment:
I refactor things a bit in when I was considering @rshriram comments about using RWMutex. As a result all of the caching stuff (get/update) is moved into helper functions which are guarded by disableCache flag which should address your concern about wasting memory.